internal/k8sinventory/watch/observer.go

```diff
@@ -0,0 +1,228 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package watch // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/k8sinventory/watch"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"go.uber.org/zap"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	apiWatch "k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/watch"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/k8sinventory"
+)
+
+const defaultResourceVersion = "1"
+
+type Config struct {
+	k8sinventory.Config
+	IncludeInitialState bool
+	Exclude             map[apiWatch.EventType]bool
+}
+
+type Observer struct {
+	config Config
+
+	client dynamic.Interface
+	logger *zap.Logger
+
+	handleWatchEventFunc func(event *apiWatch.Event)
+}
+
+func New(client dynamic.Interface, config Config, logger *zap.Logger, handleWatchEventFunc func(event *apiWatch.Event)) (*Observer, error) {
+	o := &Observer{
+		client:               client,
+		config:               config,
+		logger:               logger,
+		handleWatchEventFunc: handleWatchEventFunc,
+	}
+	return o, nil
+}
+
+func (o *Observer) Start(ctx context.Context, wg *sync.WaitGroup) chan struct{} {
+	resource := o.client.Resource(o.config.Gvr)
+	o.logger.Info("Started collecting",
+		zap.Any("gvr", o.config.Gvr),
+		zap.Any("mode", "watch"),
+		zap.Any("namespaces", o.config.Namespaces))
+
+	stopperChan := make(chan struct{})
+
+	if len(o.config.Namespaces) == 0 {
+		wg.Add(1)
+		go o.startWatch(ctx, resource, stopperChan, wg)
+	} else {
+		for _, ns := range o.config.Namespaces {
+			wg.Add(1)
+			go o.startWatch(ctx, resource.Namespace(ns), stopperChan, wg)
+		}
+	}
+
+	return stopperChan
+}
+
+func (o *Observer) startWatch(ctx context.Context, resource dynamic.ResourceInterface, stopperChan chan struct{}, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	if o.config.IncludeInitialState {
+		o.sendInitialState(ctx, resource)
+	}
+
+	watchFunc := func(options metav1.ListOptions) (apiWatch.Interface, error) {
+		options.FieldSelector = o.config.FieldSelector
+		options.LabelSelector = o.config.LabelSelector
+		return resource.Watch(ctx, options)
+	}
+
+	cancelCtx, cancel := context.WithCancel(ctx)
+
+	wait.UntilWithContext(cancelCtx, func(newCtx context.Context) {
+		resourceVersion, err := getResourceVersion(newCtx, o.config.Config, resource)
```

Comment:
In the current code, watchers have their own copy of the config:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/21380a744d44d387e60054cc17a827debc8a3274/receiver/k8sobjectsreceiver/receiver.go#L288-L310

while here, config seems to be shared, so all goroutines will operate on the same `ResourceVersion`, right?